### ISSUE TITLE
Connect to a database

### ISSUE THREAD

**Issue by jbpotonnier** [agrafix/Spock#57]
I'm trying to connect to a RethinkDB database, but I don't understand how to do it with Spock.

I created a `PoolOrConn` like this : 
```haskell
dbConn :: PoolOrConn R.RethinkDBHandle
dbConn =
  PCConn (ConnBuilder
          (R.connect "localhost" 28015 Nothing)
          R.close
          (PoolCfg 5 5 60)
         )
```
I typed my actions like this one :
```haskell
insertMessageAction :: SpockAction R.RethinkDBHandle session state ()
```
Since I use a websocket app (named `wsapp`), my main function looks like this :
```haskell
main :: IO ()
main = do
  app <- spockAsApp $ spockT _g $ appMiddleware >> appRoutes
  Warp.runSettings
        (Warp.setPort 3000 Warp.defaultSettings)
        (WaiWS.websocketsOr WS.defaultConnectionOptions wsapp app)
```
It looks I have to put something in he `_g` hole. Probably it has to do with monad transformers, which sadly I don't fully understand yet.

Could you provide some help?

**Comment by agrafix** [comment]
If you are working with 'SpockAction' you'll have to use 'spock' instead of 'spockT'. But in the example above you can fill the typehole with 'id'.

**Comment by jbpotonnier** [comment]
Thanks, it works great with `spock`.
I'm not advanced enough to find out just with the type how to configure a database pool.
It could be nice to have an example in the doc or a blog post.

**Comment by agrafix** [comment]
There's the example project here: https://github.com/agrafix/funblog . Basically `spock` and `SpockM`/`SpockAction` are specialized versions of `spockT`, `SpockT` and `ActionT`. The later only provides basic web framework functionality like routing and request handling, while `spock` and co. provide all the goodies like database connections, sessions, etc. You are right, maybe I should make that more clear in the documentation. Do you have a suggestion?

**Comment by wiria** [comment]
I don't know if this one helps you or not, but it's very basic. 
There're only 3 endpoints: 
/ which simply returns "Hello world"
string "i" followed by todo's description for the insert and
string "d" followed by todo's ID for the delete.

https://github.com/wiria/simplest-spock-SQLite-sample/blob/master/src/Main.hs

**Comment by agrafix** [comment]
Thanks for the example, @wiria , but there's some trouble with your code. You're establishing a new connection to your (sqlite) database with every request ( https://github.com/wiria/simplest-spock-SQLite-sample/blob/bfbbdd76ca5f941a2dda8bff22717936ca34ee40/src/Main.hs#L57 ). This is an anti-pattern and should be avoided. That's why connection pooling is built in with Spock and @jbpotonnier approach from above is the way to go.

**Comment by wiria** [comment]
Ah, you're correct. Thanks for pointing it out.
I was trying to show that it's quite easy to connect and doing DB operation from a Spock app.
And yes, I should modify my sample to avoid confusion.

Btw, as it has been pointed out (from the funblog sample), the correct way should be something like this:
https://github.com/agrafix/funblog/blob/master/src/Web/Blog.hs#L62 and apply [runSQL] (https://github.com/agrafix/funblog/blob/master/src/Web/Utils.hs#L28) function with that pool

**Comment by jbpotonnier** [comment]
@agrafix I didn't see the funblog example in the first place. I was rather searching in the hackage documentation.
The tutorial is using spockT function, so I thought it was the way to go.
My suggestion would be to add a canonical example in the hackage documentation, and maybe, extract the transformers in another module.

**Comment by agrafix** [comment]
Moving the transformers sounds like a good idea!

**Comment by cgag** [comment]
I have a small example project I'm working on that I intend to blog around at some point, but its a small example that hits on a fair number of features and might be useful for osme people as is: https://github.com/cgag/spock-starter

**Comment by agrafix** [comment]
Thanks @cgag , looks good and simple! If you like you can send a pull request to have it included in the readme.

**Comment by agrafix** [comment]
Closed in favor of #62
